runtime/sema/check_transaction_declaration.go

```diff
@@ -248,36 +302,157 @@ func (checker *Checker) declareTransactionDeclaration(declaration *ast.Transacti
 		transactionType.Parameters = checker.parameters(declaration.ParameterList)
 	}
 
+	members, fields, origins := checker.compositeFieldMembersAndOrigins(
+		transactionType,
+		declaration.DeclarationKind(),
+		declaration.Fields,
+	)
+
+	transactionType.Members = members
+	transactionType.Fields = fields
+	if checker.PositionInfo != nil {
+		checker.PositionInfo.recordMemberOrigins(transactionType, origins)
+	}
+
+	if declaration.Prepare != nil {
+		parameterList := declaration.Prepare.FunctionDeclaration.ParameterList
+		transactionType.PrepareParameters = checker.parameters(parameterList)
+	}
+
+	for _, roleDeclaration := range declaration.Roles {
+		transactionRoleType := checker.transactionRoleType(roleDeclaration)
+		checker.Elaboration.SetTransactionRoleDeclarationType(roleDeclaration, transactionRoleType)
+
+		// Ensure roles and fields do not clash
+		roleName := roleDeclaration.Identifier.Identifier
+		if member, ok := members.Get(roleName); ok {
+
+			identifierRange := ast.NewRangeFromPositioned(
+				checker.memoryGauge,
+				roleDeclaration.Identifier,
+			)
+
+			if _, ok := member.TypeAnnotation.Type.(*TransactionRoleType); ok {
```

Comment:
not really a big deal: we can also check for `member.DeclarationKind == common.DeclarationKindTransactionRole` instead of the type cast?

Reply:
That works too, changed in bb6bebaa137b4c24ccf78c2ac526db9345d565c0